src/spatialdata/dataloader/graph_loader.py

```diff
@@ -29,104 +46,90 @@ def build_graph(  # TODO: add rings for grids
-                centroids = np.array(centroids)
-                dd = tree.sparse_distance_matrix(
-                    tree,
-                    max_distance=np.sqrt(
-                        (np.max(centroids[:, 0]) - np.min(centroids[:, 0])) ** 2
-                        + (np.max(centroids[:, 1]) - np.min(centroids[:, 1])) ** 2
-                    ),
-                ).toarray()
-                np.fill_diagonal(dd, np.nan)
-                dd = dd[~np.isnan(dd)].reshape(dd.shape[0], dd.shape[1] - 1)
-                max_distance = np.percentile(dd, percentile)
-                d_tree, idx = tree.query(centroids, k=k, distance_upper_bound=max_distance)
-            else:
-                d_tree, idx = tree.query(centroids, k=k)
+            d_tree, idx = tree.query(centroids, k=k)
 
         idx = idx[:, 1:]
         if idx.size == 0:
-            return
+            return np.array([[]])
         point_numbers = np.arange(len(centroids))
+        assert isinstance(k, int)
         point_numbers = np.repeat(point_numbers, k - 1)
         idx_flatten = idx.flatten()
         edge_index = np.vstack((point_numbers, idx_flatten)).T
-
         # to delete the edges longer than max_distance
         # check for self edges
     elif method == "radius":
-        assert k is None
-        assert max_distance is not None or percentile is not None
-        if not use_centroids:
-            logger.warning("method radius is only available for centroids, automatically set use_centroids=True.")
-        gdf_t = gdf.centroid
-        centroids = gdf_t.apply(lambda g: [g.x, g.y]).tolist()
+        gdf_c = gdf.centroid
+        centroids = gdf_c.apply(lambda g: [g.x, g.y]).tolist()
         tree = KDTree(centroids)
         if max_distance is not None:
-            if percentile is not None:
-                logger.warning("both percentile and max_distance are set, will only use max_distance.")
             neigh_list = tree.query_ball_tree(tree, r=max_distance)
         else:
-            logger.warning("computing percentile could take long for large datasets.")
-            centroids = np.array(centroids)
-            dd = tree.sparse_distance_matrix(
-                tree,
-                max_distance=np.sqrt(
-                    (np.max(centroids[:, 0]) - np.min(centroids[:, 0])) ** 2
-                    + (np.max(centroids[:, 1]) - np.min(centroids[:, 1])) ** 2
-                ),
-            ).toarray()
-            np.fill_diagonal(dd, np.nan)
-            dd = dd[~np.isnan(dd)].reshape(dd.shape[0], dd.shape[1] - 1)
-            max_distance = np.percentile(dd, percentile)
+            assert isinstance(percentile, int)
+            max_distance = _compute_percentile(tree, percentile, centroids)
             neigh_list = tree.query_ball_tree(tree, r=max_distance)
-        edge_index = []
+        edge_index_list: list[list[int]] = []
         for sublist_index, sublist in enumerate(neigh_list):
-            for index in sublist:
```

Comment:
@EliHei2 I removed this line. `index` was never used. Please if now it's correct.